# (Project 1)

# working on Weather Data by using pandas dataframe

The weather data set is time series data set with per-hour information about the weather conditions at a particular lacation. It records Temperature. Dew Point,Temperature,Relative Humidity,visibility,Pressure and Conditions. 


In [5]:
import pandas as pd

In [15]:
#ready-code from IBM watson studio to read any loaded file 
import os, types
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.
client_89964c8d3f9f498cba0461a952edae93 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='FiDOa7PobmyuC_jhqU2zWgBMIP9jrskGpyRbERg45_aL',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3.private.eu.cloud-object-storage.appdomain.cloud')

body = client_89964c8d3f9f498cba0461a952edae93.get_object(Bucket='dataanalyticsprojects-donotdelete-pr-dfxqqqpn3k5k03',Key='1. Weather Data.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df = pd.read_csv(body)
df.head()


Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [16]:
df #It shows dataframe with first and last five rows

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

>>>>Just to take overview of dataframe

Before Analysis let's explore The data

In [17]:
df.head() #shows dataframe with first 5 rows here 5 rows are default we can show any no. of rows by giving argument in ()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [19]:
df.shape # To show total no. of rows and columns

(8784, 8)

In [20]:
df.index # shows index in dataframe with starts and end indices

RangeIndex(start=0, stop=8784, step=1)

In [21]:
df.columns # shows columns name in the dataframe

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

In [22]:
df.dtypes #shows data types of each column in dataframe

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

In [28]:
df['Weather'].unique() # shows unique rows in particular column

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

In [29]:
df.nunique() # To see no. of unique values in each column in dataframe We can use this on whole dataframe unlike unique() command

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

In [30]:
df.count() # show total no. of non-null vlaues in dataframe .We can also use this for columns

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

>>>> There is no null value in the given dataframe

In [33]:
df['Weather'].value_counts() # shows unique values with their counts . can apply on columns. If we apply it on whole data frame it would show unique values in each colmns

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

In [38]:
df.value_counts()

Date/Time        Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  Visibility_km  Press_kPa  Weather     
1/1/2012 0:00    -1.8    -3.9              86         4                8.0            101.24     Fog             1
6/1/2012 12:00    19.3    3.3              35         20               48.3           101.32     Cloudy          1
5/9/2012 7:00     14.3    12.5             89         15               4.8            100.12     Fog             1
5/9/2012 8:00     14.3    12.3             88         17               6.4            100.12     Fog             1
5/9/2012 9:00     14.0    12.3             89         9                4.0            100.10     Drizzle,Fog     1
                                                                                                                ..
12/8/2012 3:00    2.1    -1.5              77         6                25.0           101.18     Cloudy          1
12/8/2012 2:00    2.0    -1.9              75         7                25.0          

In [39]:
df.info() # shows index ranges ,data columns,data types 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


### 1)To find all the unique 'wind speed' values in the data

In [40]:
df['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

### 2) Find the number of times when the 'weather is exactly clear' 

In [49]:
#METHOD_1 by using value_counts()
df.Weather.value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

In [51]:
#METHOD_2 by using filtering
df[df.Weather== 'Clear']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [52]:
#METHOD_3 by using groupby()
df.groupby('Weather').get_group('Clear')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

#### answer>>>1326

### 3) find the number of times when the 'wind speed was exactly 4 km/h'

In [56]:
df['Wind Speed_km/h'].value_counts()

9     830
11    791
13    735
15    719
7     677
17    666
19    616
6     609
20    496
4     474
22    439
24    374
0     309
26    242
28    205
30    161
32    139
33     85
35     53
37     45
39     24
41     22
44     14
48     13
43     13
46     11
52      7
57      5
50      4
2       2
70      1
54      1
83      1
63      1
Name: Wind Speed_km/h, dtype: int64

In [57]:
# using filtering
df[df['Wind Speed_km/h']== 4]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

#### answer>>> 474

### 4)Find out all the Null Values in the data.

In [60]:
#using isnull()
df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

In [61]:
#using notnull()
df.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

#### answer>>> 0

### 5)Rename the column 'weather' of datafrme to 'weather condition'

In [63]:
df.rename(columns ={'Weather': 'Weather Condition'},inplace=True) # Here we use inplace=True to change column name in original data frame othewise it would rename column name temporary

In [71]:
df

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

### 6)what is mean 'visibility'?

In [66]:
# df['Visibility_km'].mean() ### we will use this if column names contains space . otherwise we will use folloing syntax
df.Visibility_km.mean()

27.664446721311478

### 7)what is standard deviation of 'pressure' in data? 

In [67]:
df.Press_kPa.std()

0.8440047459486483

### 8)what is variance of ' Relative Humidity' in this data?

In [69]:
df['Rel Hum_%'].var()

286.24855019850196

### 9)Find all instances when snow was recorded

In [72]:
# by value_counts()
df['Weather Condition'].value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

In [73]:
# by using filter
df[df['Weather Condition']== 'Snow']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
55             25.0     100.95              Snow  
84             24.1     101.25              Snow  
86             19.3     100.97              Snow  
87              9.7     100.89              Snow  
88             19.3     100.79              Snow  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[390 rows x 8 columns]

In [77]:
# str.contains
df[df['Weather Condition'].str.contains('Snow')].head(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
41   1/2/2012 17:00    -2.1              -9.5         57               22   
44   1/2/2012 20:00    -5.6             -13.4         54               24   
45   1/2/2012 21:00    -5.8             -12.8         58               26   
47   1/2/2012 23:00    -7.4             -14.1         59               17   
48    1/3/2012 0:00    -9.0             -16.0         57               28   
50    1/3/2012 2:00   -10.5             -15.8         65               22   
51    1/3/2012 3:00   -11.3             -18.7         54               33   
53    1/3/2012 5:00   -12.9             -19.1         60               22   
54    1/3/2012 6:00   -13.3             -19.3         61               19   
55    1/3/2012 7:00   -14.0             -19.5         63               19   
84   1/4/2012 12:00   -13.7             -21.7         51               11   
86   1/4/2012 14:00   -11.3             -19.0         53                7   
87   1/4/2012 15:00   -10.2             -16.3         61               11   
88   1/4/2012 16:00    -9.4             -15.5         61               13   
89   1/4/2012 17:00    -8.9             -13.2         71                9   
90   1/4/2012 18:00    -8.9             -12.6         75               11   
91   1/4/2012 19:00    -8.4             -12.7         71                9   
92   1/4/2012 20:00    -7.8             -12.1         71                9   
93   1/4/2012 21:00    -7.6             -11.6         73                7   
94   1/4/2012 22:00    -9.5             -12.7         77                6   
95   1/4/2012 23:00    -9.6             -12.6         79                6   
96    1/5/2012 0:00    -8.8             -11.7         79                4   
97    1/5/2012 1:00    -7.5             -10.2         81                0   
98    1/5/2012 2:00    -5.4              -8.3         80                9   
99    1/5/2012 3:00    -5.0              -7.7         81               11   
100   1/5/2012 4:00    -4.1              -6.5         83                9   
101   1/5/2012 5:00    -7.0              -9.5         82                4   
102   1/5/2012 6:00    -8.2             -10.7         82                6   
103   1/5/2012 7:00    -7.1              -9.7         82                9   
104   1/5/2012 8:00    -6.1              -9.1         79               11   
123   1/6/2012 3:00   -10.6             -16.0         64                0   
124   1/6/2012 4:00   -11.3             -16.1         68               15   
125   1/6/2012 5:00   -11.8             -16.0         71               19   
126   1/6/2012 6:00   -12.0             -16.2         71               22   
127   1/6/2012 7:00   -14.4             -16.3         85               22   
128   1/6/2012 8:00   -12.3             -16.2         73               24   
129   1/6/2012 9:00   -12.5             -16.7         71               26   
130  1/6/2012 10:00   -12.3             -16.3         72               28   
131  1/6/2012 11:00   -12.0             -16.0         72               17   
132  1/6/2012 12:00   -11.7             -15.4         74               28   
133  1/6/2012 13:00   -11.9             -15.6         74               20   
134  1/6/2012 14:00   -11.2             -14.8         75               19   
135  1/6/2012 15:00   -11.5             -14.4         79               19   
136  1/6/2012 16:00   -11.6             -14.7         78               19   
137  1/6/2012 17:00   -11.2             -14.3         78               17   
158  1/7/2012 14:00    -4.4              -6.9         83                6   
159  1/7/2012 15:00    -3.7              -6.1         83               11   
160  1/7/2012 16:00    -3.6              -5.9         84                0   
161  1/7/2012 17:00    -3.1              -5.4         84               13   
162  1/7/2012 18:00    -3.2              -5.3         85                6   

     Visibility_km  Press_kPa Weather Condition  
41            25.0      99.66   

#### answer >>> 390

### 10)Find all instances when wind speed is 'above 24 ' and 'visibility' is 'above 25'

In [80]:
df[(df['Wind Speed_km/h']>24) & (df['Visibility_km']==25)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa Weather Condition  
23             25.0      99.31            Cloudy  
24             25.0      99.26      Rain Showers  
25             25.0      99.26            Cloudy  
26             25.0      99.26     Mostly Cloudy  
27             25.0      99.30     Mostly Cloudy  
...             ...        ...               ...  
8705           25.0     101.34      Mainly Clear  
8753           25.0     101.26      Mainly Clear  
8755           25.0     101.47      Mainly Clear  
8759           25.0     101.52     Mostly Cloudy  
8760           25.0     101.51            Cloudy  

[308 rows x 8 columns]

### 11)what is mean value of each column againts each 'weather condition'?

In [82]:
df.groupby('Weather Condition').mean()

Temp_C  Dew Point Temp_C  \
Weather Condition                                                      
Clear                                     6.825716          0.089367   
Cloudy                                    7.970544          2.375810   
Drizzle                                   7.353659          5.504878   
Drizzle,Fog                               8.067500          7.033750   
Drizzle,Ice Pellets,Fog                   0.400000         -0.700000   
Drizzle,Snow                              1.050000          0.150000   
Drizzle,Snow,Fog                          0.693333          0.120000   
Fog                                       4.303333          3.159333   
Freezing Drizzle                         -5.657143         -8.000000   
Freezing Drizzle,Fog                     -2.533333         -4.183333   
Freezing Drizzle,Haze                    -5.433333         -8.000000   
Freezing Drizzle,Snow                    -5.109091         -7.072727   
Freezing Fog                             -7.575000         -9.250000   
Freezing Rain                            -3.885714         -6.078571   
Freezing Rain,Fog                        -2.225000         -3.750000   
Freezing Rain,Haze                       -4.900000         -7.450000   
Freezing Rain,Ice Pellets,Fog            -2.600000         -3.700000   
Freezing Rain,Snow Grains                -5.000000         -7.300000   
Haze                                     -0.200000         -2.975000   
Mainly Clear                             12.558927          4.581671   
Moderate Rain,Fog                         1.700000          0.800000   
Moderate Snow                            -5.525000         -7.250000   
Moderate Snow,Blowing Snow               -5.450000         -6.500000   
Mostly Cloudy                            10.574287          3.131174   
Rain                                      9.786275          7.042810   
Rain Showers                             13.722340          9.187766   
Rain Showers,Fog                         12.800000         12.100000   
Rain Showers,Snow Showers                 2.150000         -1.500000   
Rain,Fog                                  8.273276          7.219828   
Rain,Haze                                 4.633333          2.066667   
Rain,Ice Pellets                          0.600000         -0.600000   
Rain,Snow                                 1.055556         -0.566667   
Rain,Snow Grains                          1.900000         -2.100000   
Rain,Snow,Fog                             0.800000          0.300000   
Rain,Snow,Ice Pellets                     1.100000         -0.175000   
Snow                                     -4.524103         -7.623333   
Snow Pellets                              0.700000         -6.400000   
Snow Showers                             -3.506667         -7.866667   
Snow Showers,Fog                        -10.675000        -11.900000   
Snow,Blowing Snow                        -5.410526         -7.621053   
Snow,Fog                                 -5.075676         -6.364865   
Snow,Haze                                -4.020000         -6.860000   
Snow,Ice Pellets                         -1.883333         -3.666667   
Thunderstorms                            24.150000         19.750000   
Thunderstorms,Heavy Rain Showers         10.900000          9.000000   
Thunderstorms,Moderate Rain Showers,Fog  19.600000         18.500000   
Thunderstorms,Rain                       20.433333         18.533333   
Thunderstorms,Rain Showers               20.037500         17.618750   
Thunderstorms,Rain Showers,Fog           21.600000         18.700000   
Thunderstorms,Rain,Fog                   20.600000         18.600000   

                                         Rel Hum_%  Wind Speed_km/h  \
Weather Condition                                                     
Clear                                    64.497738        10.557315   
Cloudy                                   69.592593        16.127315   
Drizzle        

### 12)what is minimum and maximum value of each column againts each 'weather condition'? 

In [87]:
df.groupby('Weather Condition').min()

Date/Time  Temp_C  \
Weather Condition                                                   
Clear                                      1/11/2012 1:00   -23.3   
Cloudy                                     1/1/2012 17:00   -21.4   
Drizzle                                   1/23/2012 21:00     1.1   
Drizzle,Fog                               1/23/2012 20:00     0.0   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/17/2012 15:00     0.9   
Drizzle,Snow,Fog                         12/18/2012 21:00     0.3   
Fog                                         1/1/2012 0:00   -16.0   
Freezing Drizzle                          1/13/2012 10:00    -9.0   
Freezing Drizzle,Fog                        1/1/2012 2:00    -6.4   
Freezing Drizzle,Haze                      2/1/2012 11:00    -5.8   
Freezing Drizzle,Snow                      1/13/2012 3:00    -8.3   
Freezing Fog                               1/22/2012 6:00   -19.0   
Freezing Rain                             1/13/2012 11:00    -6.5   
Freezing Rain,Fog                         1/17/2012 23:00    -6.1   
Freezing Rain,Haze                         2/1/2012 14:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      1/22/2012 12:00   -11.5   
Mainly Clear                              1/10/2012 11:00   -22.8   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             1/12/2012 15:00    -6.3   
Moderate Snow,Blowing Snow               12/27/2012 10:00    -5.5   
Mostly Cloudy                              1/1/2012 16:00   -23.2   
Rain                                       1/1/2012 18:00     0.3   
Rain Showers                               1/1/2012 22:00     1.6   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                  11/4/2012 8:00     2.1   
Rain,Fog                                  1/23/2012 18:00     0.0   
Rain,Haze                                  3/13/2012 7:00     4.0   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  1/10/2012 5:00     0.6   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 1:00     0.9   
Snow                                       1/10/2012 1:00   -16.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               1/12/2012 7:00   -13.3   
Snow Showers,Fog                          12/26/2012 9:00   -11.3   
Snow,Blowing Snow                         1/13/2012 21:00   -12.0   
Snow,Fog                                 12/16/2012 15:00   -10.1   
Snow,Haze                                  2/1/2012 17:00    -4.3   
Snow,Ice Pellets                          12/10/2012 3:00    -4.3   
Thunderstorms                              7/16/2012 1:00    21.6   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        5/25/2012 20:00    19.4   
Thunderstorms,Rain Showers                5/29/2012 16:00    11.0   
Thunderstorms,Rain Showers,Fog             6/29/2012 3:00    19.5   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                               -28.5         20   
Cloudy                                              -26.8         18   
Drizzle                                              -0.2         74   
Drizzle,Fog                                          -1.6         85   
Drizzle,Ice Pellets,Fog    

In [88]:
df.groupby('Weather Condition').max()

Date/Time  Temp_C  \
Weather Condition                                                   
Clear                                       9/9/2012 5:00    32.8   
Cloudy                                     9/9/2012 23:00    30.5   
Drizzle                                    9/30/2012 3:00    18.8   
Drizzle,Fog                                9/30/2012 2:00    19.9   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/19/2012 18:00     1.2   
Drizzle,Snow,Fog                          12/22/2012 3:00     1.1   
Fog                                        9/22/2012 0:00    20.8   
Freezing Drizzle                            2/1/2012 5:00    -2.3   
Freezing Drizzle,Fog                      12/10/2012 5:00    -0.3   
Freezing Drizzle,Haze                      2/1/2012 13:00    -5.0   
Freezing Drizzle,Snow                      3/2/2012 12:00    -3.3   
Freezing Fog                               3/17/2012 6:00    -0.1   
Freezing Rain                               2/1/2012 7:00     0.3   
Freezing Rain,Fog                         12/17/2012 1:00     0.1   
Freezing Rain,Haze                         2/1/2012 15:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      3/13/2012 23:00    14.1   
Mainly Clear                                9/9/2012 9:00    33.0   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             12/27/2012 9:00    -4.9   
Moderate Snow,Blowing Snow               12/27/2012 12:00    -5.4   
Mostly Cloudy                               9/9/2012 2:00    32.4   
Rain                                        9/5/2012 2:00    22.8   
Rain Showers                               9/8/2012 16:00    26.4   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                 12/5/2012 10:00     2.2   
Rain,Fog                                  9/30/2012 23:00    21.7   
Rain,Haze                                  3/13/2012 9:00     5.5   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  4/23/2012 3:00     1.7   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 5:00     1.3   
Snow                                       4/27/2012 9:00     3.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               3/4/2012 21:00     2.9   
Snow Showers,Fog                         12/29/2012 13:00   -10.0   
Snow,Blowing Snow                          2/25/2012 9:00    -1.4   
Snow,Fog                                  3/14/2012 19:00     1.1   
Snow,Haze                                  2/1/2012 21:00    -3.6   
Snow,Ice Pellets                            3/3/2012 4:00     0.8   
Thunderstorms                              7/4/2012 16:00    26.7   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        7/23/2012 18:00    21.3   
Thunderstorms,Rain Showers                  9/8/2012 4:00    25.5   
Thunderstorms,Rain Showers,Fog            7/31/2012 20:00    22.9   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                                20.4         99   
Cloudy                                               22.6         99   
Drizzle                                              17.7         96   
Drizzle,Fog                                          19.1        100   
Drizzle,Ice Pellets,Fog    

### 13)show all the records where weather condition is 'Fog'

In [89]:
df[df['Weather Condition']=='Fog']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather Condition  
0               8.0     101.24               Fog  
1               8.0     101.24               Fog  
4               4.8     101.23               Fog  
5               6.4     101.27               Fog  
6               6.4     101.29               Fog  
...             ...        ...               ...  
8716            9.7     101.25               Fog  
8717            6.4     101.25               Fog  
8718            9.7     101.25               Fog  
8719            8.0     101.22               Fog  
8722            6.4     101.15               Fog  

[150 rows x 8 columns]

### 14)Find all the instances where ,'weather' is 'clear' or 'visibility' is above 40

In [94]:
df[(df['Weather Condition'] == 'Clear') | (df['Visibility_km']>40)].head(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67    1/3/2012 19:00   -16.9             -24.8         50               24   
106   1/5/2012 10:00    -6.0             -10.0         73               17   
107   1/5/2012 11:00    -5.6             -10.2         70               22   
108   1/5/2012 12:00    -4.7              -9.6         69               20   
109   1/5/2012 13:00    -4.4              -9.7         66               26   
110   1/5/2012 14:00    -5.1             -10.7         65               22   
111   1/5/2012 15:00    -4.3             -12.0         55               26   
114   1/5/2012 18:00    -7.1             -14.4         56               11   
115   1/5/2012 19:00    -9.2             -15.4         61                7   
116   1/5/2012 20:00    -9.8             -15.7         62                9   
117   1/5/2012 21:00    -9.0             -14.8         63               13   
183   1/8/2012 15:00    -6.6             -12.9         61               20   
241   1/11/2012 1:00   -10.7             -17.8         56               17   
242   1/11/2012 2:00   -12.0             -18.9         56               19   
243   1/11/2012 3:00   -12.7             -19.4         57               19   
244   1/11/2012 4:00   -13.4             -20.1         57               17   
324  1/14/2012 12:00   -17.5             -23.8         58               20   
325  1/14/2012 13:00   -17.1             -24.1         55               17   
326  1/14/2012 14:00   -16.7             -23.4         56               17   
327  1/14/2012 15:00   -16.7             -23.4         56               22   
344   1/15/2012 8:00   -23.3             -28.5         62                7   
345   1/15/2012 9:00   -22.2             -27.8         60                9   
346  1/15/2012 10:00   -20.6             -26.8         58                9   
347  1/15/2012 11:00   -19.3             -26.1         55                9   
348  1/15/2012 12:00   -18.0             -25.5         52               13   
349  1/15/2012 13:00   -16.8             -24.2         53               20   
350  1/15/2012 14:00   -16.0             -23.4         53               26   
351  1/15/2012 15:00   -15.4             -22.8         53               24   
352  1/15/2012 16:00   -15.1             -22.8         52               24   
353  1/15/2012 17:00   -16.2             -23.2         55               15   
354  1/15/2012 18:00   -16.3             -22.9         57               17   
355  1/15/2012 19:00   -16.3             -22.7         58               20   
356  1/15/2012 20:00   -16.7             -22.4         61               17   
357  1/15/2012 21:00   -16.9             -22.2         63               15   
358  1/15/2012 22:00   -16.9             -21.7         66               17   
359  1/15/2012 23:00   -16.9             -21.4         68               15   
360   1/16/2012 0:00   -17.1             -21.9         66               15   
361   1/16/2012 1:00   -17.8             -22.7         65                6   
362   1/16/2012 2:00   -18.5             -23.0         68                6   
363   1/16/2012 3:00   -19.2             -23.8         67                6   
364   1/16/2012 4:00   -18.7             -23.5         66                0   
365   1/16/2012 5:00   -19.1             -23.3         69                0   
366   1/16/2012 6:00   -18.7             -23.2         68                0   
367   1/16/2012 7:00   -17.9             -21.5         73                4   
368   1/16/2012 8:00   -17.9             -22.6         67                9   
369   1/16/2012 9:00   -14.4             -20.7         59               22   
370  1/16/2012 10:00   -13.6             -18.7         65               19   
371  1/16/2012 11:00   -12.7             -17.7         66               20   
372  1/16/2012 12:00   -10.7             -16.9         60               19   
373  1/16/2012 13:00   -10.1             -16.1         61               13   

     Visibility_km  Press_kPa We

### 15)Find all instances when 'weather is clear' and 'relative Humidity is greater than 50' or 'visibility is above 40

In [97]:
df[(df['Weather Condition']== 'Clear')&(df['Rel Hum_%']>50) | (df['Visibility_km']>40) ].head(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
106   1/5/2012 10:00    -6.0             -10.0         73               17   
107   1/5/2012 11:00    -5.6             -10.2         70               22   
108   1/5/2012 12:00    -4.7              -9.6         69               20   
109   1/5/2012 13:00    -4.4              -9.7         66               26   
110   1/5/2012 14:00    -5.1             -10.7         65               22   
111   1/5/2012 15:00    -4.3             -12.0         55               26   
114   1/5/2012 18:00    -7.1             -14.4         56               11   
115   1/5/2012 19:00    -9.2             -15.4         61                7   
116   1/5/2012 20:00    -9.8             -15.7         62                9   
117   1/5/2012 21:00    -9.0             -14.8         63               13   
183   1/8/2012 15:00    -6.6             -12.9         61               20   
241   1/11/2012 1:00   -10.7             -17.8         56               17   
242   1/11/2012 2:00   -12.0             -18.9         56               19   
243   1/11/2012 3:00   -12.7             -19.4         57               19   
244   1/11/2012 4:00   -13.4             -20.1         57               17   
324  1/14/2012 12:00   -17.5             -23.8         58               20   
325  1/14/2012 13:00   -17.1             -24.1         55               17   
326  1/14/2012 14:00   -16.7             -23.4         56               17   
327  1/14/2012 15:00   -16.7             -23.4         56               22   
344   1/15/2012 8:00   -23.3             -28.5         62                7   
345   1/15/2012 9:00   -22.2             -27.8         60                9   
346  1/15/2012 10:00   -20.6             -26.8         58                9   
347  1/15/2012 11:00   -19.3             -26.1         55                9   
348  1/15/2012 12:00   -18.0             -25.5         52               13   
349  1/15/2012 13:00   -16.8             -24.2         53               20   
350  1/15/2012 14:00   -16.0             -23.4         53               26   
351  1/15/2012 15:00   -15.4             -22.8         53               24   
352  1/15/2012 16:00   -15.1             -22.8         52               24   
353  1/15/2012 17:00   -16.2             -23.2         55               15   
354  1/15/2012 18:00   -16.3             -22.9         57               17   
355  1/15/2012 19:00   -16.3             -22.7         58               20   
356  1/15/2012 20:00   -16.7             -22.4         61               17   
357  1/15/2012 21:00   -16.9             -22.2         63               15   
358  1/15/2012 22:00   -16.9             -21.7         66               17   
359  1/15/2012 23:00   -16.9             -21.4         68               15   
360   1/16/2012 0:00   -17.1             -21.9         66               15   
361   1/16/2012 1:00   -17.8             -22.7         65                6   
362   1/16/2012 2:00   -18.5             -23.0         68                6   
363   1/16/2012 3:00   -19.2             -23.8         67                6   
364   1/16/2012 4:00   -18.7             -23.5         66                0   
365   1/16/2012 5:00   -19.1             -23.3         69                0   
366   1/16/2012 6:00   -18.7             -23.2         68                0   
367   1/16/2012 7:00   -17.9             -21.5         73                4   
368   1/16/2012 8:00   -17.9             -22.6         67                9   
369   1/16/2012 9:00   -14.4             -20.7         59               22   
370  1/16/2012 10:00   -13.6             -18.7         65               19   
371  1/16/2012 11:00   -12.7             -17.7         66               20   
372  1/16/2012 12:00   -10.7             -16.9         60               19   
373  1/16/2012 13:00   -10.1             -16.1         61               13   
374  1/16/2012 14:00    -9.0             -13.8         68               13   

     Visibility_km  Press_kPa We